flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/TaskStateAssignment.java

```diff
@@ -166,32 +169,47 @@ public OperatorSubtaskState getSubtaskState(OperatorInstanceID instanceID) {
                                 instanceID,
                                 inputOperatorID,
                                 getUpstreamAssignments(),
-                                assignment -> assignment.outputSubtaskMappings,
+                                assignment ->
+                                        assignment.outputSubtaskMappings.get(
+                                                getAssignmentIndex(
+                                                        assignment.getDownstreamAssignments(),
+                                                        this)),
                                 assignment ->
                                         assignment.getOutputMapping(
-                                                Arrays.asList(assignment.getDownstreamAssignments())
-                                                        .indexOf(this)),
+                                                getAssignmentIndex(
+                                                        assignment.getDownstreamAssignments(),
+                                                        this)),
```

Comment:
Didn't you had a nice idea on how to reduce this monster? 
One option could be to pass a parameter through the lambda to `getOutputMapping` indicating if it should lazy initialize or not.

Reply:
I thought about different ways, but none of the ideas I had, looked so much better that I thought about implementing it. 

I added your suggestion (minus that the check is in the lambda rather than `getOutputMapping`). Take a look, not sure if it is much better :(